### DISCUSSION TITLE
PICO-MicroPython and USB-CDC time-out

### DISCUSSION THREAD

**Discussion by Filip-Declercq** [hathach/tinyusb#1530] [Q&A] [answered]
Hello. I've got a project using an RPi PICO running MicroPython. MicroPython exposes its standard I/O over USB, using a virtual com port that is created using the tinyusb library. Now I see that this CDC serial port turns irresponsive and shows time-out errors. I use the hardware uart quite intensive to communicate with a modem module. The RP2 port of MicroPython has implemented an interrupt driven ring buffer to serve the uart stream library. I guess that having a near constant stream of uart i/o starves the TinyUSB tud_task(). Result: the REPL that is availble over USB on the PICO, stops being responsive because the USB CDC device is no longer available. Time-out errors occur. If I add extra "sleep" statements now and then, I get the system to stay resposive, but I'm not sure how long to wait, and I don't think this is a clean solution.

Any suggestions/tips on how to cope with this problem? 
Is there a directive of the minimal repetition frequency/max delay of the tud_task?

Or should I better move to C? Anyone who could confirm that this is a good marriage: RP2 (PICO)/FreeRTOS (with CDC through TinyUSB)/uart/multithreading using the two RP2040 cores?

**Comment by hathach** [comment, marked as answer]
for micropython integration issue, you should ask this on micropython forum.

**Reply by Filip-Declercq** [reply]
Hi, I understand you direct me to the MicroPython forum, but I guess I'm on the right forum for this question:
Is there a directive of the minimal repetition frequency/max delay of the tud_task?
E.g. if I want to use TinyUSB in C

**Reply by pigrew** [reply]
From a USB-point of view, the shortest deadline is 50ms for completing the setup or status stage of standard device requests without a data stage (USB spec 2.0, section 9.2.6.4), so sometime shorter than that should be the deadline. Standard requests with a data stage have a 500ms deadline, and "normal" class requests have a 5-second deadline for processing (unless the class itself has a tighter deadline). TinyUSB doesn't have any requirements that are any tighter than than the USB spec itself.
